[체스 - 1, 2단계] 아토(이혜린) 미션 제출합니다.

hyxrxn opened the pull request:
호돌 안녕하세요! BE 6기 **아토**입니다.
이번 미션은 일단 요구사항을 충족하는 프로그램을 만드는 것을 가장 큰 목표로 삼았습니다!
목표 달성은 했으나 쏟은 시간 치고는 고치고 싶은 부분이 아직 많네요🤣
구현 간단하게 설명 한 번 하겠습니다!

----------------------
### 1. 구현 목표에 넣었으나 아직 시도하지 못한 부분
   - 일단 2단계 구현하며 1차 목표는 정말 단순하게 '이동' 이었습니다. 하지만 시각에 따라 이동에 포함된다고 할 수 있는 부분이 두 가지가 더 있었습니다. 첫 번째는 흰색 말을 시작으로 한 번씩 번갈아가며 이동하는지 확인하는 것이었고, 두 번째는 이동 목적지에 상대편의 말이 있다면 잡는 것이었습니다. 그런데 1차 목표를 달성하고 리팩토링을 하다 보니 더 이상 구현하기는 어려운 시간이 되어 달성은 하지 못했습니다. 코드 읽거나 실행해보실 때 참고하세요!
   - 추가로, 시작 멘트 출력 시 커맨드 키(start, end, move)를 함께 출력하는 부분이 있습니다. 이 부분 또한 페어와 함께 분리하자는 이야기를 나누었으나 아직 분리되지 않았습니다. 또한 입력이 잘못되면 바로 종료됩니다!

### 2. 말이 이동할 수 있는 방법 분리
   - 칸 제한 없이 끝까지 이동할 수 있는 방법은 continuous로 추상화하고 한 칸만 이동할 수 있는 방법은 discrete로 추상화했습니다. 처음 움직이는 폰이 두 칸 이동하는 것은 따로 분리하지 않았습니다! 사실 아직 겹치는 코드가 좀 있다고 생각하긴 합니다만, 추상화를 더 하면 오히려 복잡도만 올라가는 결과를 낳을까 우려되어 일단 그대로 두었습니다.

### 3. 이동 & 출력 관련 로직
  - 각 말이 본인의 종류에 따라 가능한 이동 방법들을 들고 있습니다. 그 방법을 이용해 목적지가 갈 수 있는 칸인지를 판단하고, 갈 수 있다면 어딜 거쳐 가는지 보드에게 알려줍니다. 보드는 거쳐가는 칸을 받아 사이에 말이 있는지 확인하고, 없다면 그 말을 이동시킵니다. 위치는 요구사항에 따라 a~h, 1~8로 표현되고, 보드가 위치에 따른 말을 알고 있습니다. 출력 시엔 보드가 칸별로 어떤 말이 있는지 알려준 결과물을 컨트롤러가 뷰에게 주고, 뷰는 칸별로 확인하며 말이 없다면 `.`을 출력합니다.

-----------
많은 의견과 질문 환영합니다!
프리코스 때 자바를 처음 접해봐서 아직 모르는 것이 많습니다.
제 의도가 잘 보이지 않거나, 비효율적인 것 같은 부분이 있다면 얼마든지 알려주세요.
좋은 하루 보내시길 바라요!
감사합니다~~~😊

hyxrxn committed:
docs: 기능 요구 사항 정리

hyxrxn committed:
feat(King): 검정 팀 여부 확인

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
docs: '팀 결정 여부 확인' 문서에 반영

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Position): 각 위치 구분

- 가로 위치와 세로 위치 저장
- 가로 위치와 세로 위치가 같으면, 같은 객체로 판단

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Board): 해당 위치의 말 확인

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(BoardFactory): 체스 말 위치 초기화

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(domain): 패키지 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(ChessGame): 초기 체스 보드 출력

- 대문자는 검정말, 소문자는 흰말로 출력

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(OutputView): 메서드 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(ChessGame): 메서드 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(OutputView): 불필요한 공백 제거

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(InputView): 커멘드 입력

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(ChessGame): 종료 커멘드 반영

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(Team): 개행 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Row): 남북 이동

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Column): 동서 이동

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Position): 다음 위치 확인

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(Rank, File): 클래스명 변경

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(MovementRule): 연속적인 움직임

- 이동 가능 여부 판단
- 해당 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(KnightMovement): 나이트 움직임

- 이동 가능 여부 판단
- 해당 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(KingMovement): 킹 움직임

- 이동 가능 여부 판단
- 해당 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Piece): 각 기물별 가능한 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Board): 시작 위치의 말을 도착 위치로 이동

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(ChessGame): 이동 커멘드 반영

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(PawnDefaultMovement): 폰 기본 움직임

- 이동 가능 여부 판단
- 해당 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
fix(Rook): 남쪽으로 이동할 수 있도록 수정

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(PawnFirstMovement): 폰 처음 움직임

- 이동 가능 여부 판단
- 해당 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style: 컨벤션에 맞추어 공백 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(KingTest): 불필요한 구문 제거

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Pawn): 가능한 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(Piece): stream을 이용하여 가독성 개선

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(ContinuousMovementRule): 패키지 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(PieceTest): 팀 확인 테스트 이동

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(Piece): 불필요한 구문 제거

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
test(Piece): 가능한 경로 파악

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(Piece): 에러 메시지 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
test(Position): 파일과 랭크 차이 계산

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(ChessGame): 메서드 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(ChessApplication): 사용자에게 에러메시지 출력 후 종료

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(DiscreteMovementRule): 공통 로직 추상 클래스로 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(ContinuousMovementRule): 불필요한 개행 제거

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(DiscreteMovementRule): 에러 메시지 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(PawnFirstMovement): 잘못된 파라미터 입력 시 에러 발생

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor: 가독성을 위해 상수 추출

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(GameCommand): 불필요한 메서드 삭제

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
feat(BoardFactory): 객체 생성을 막기 위해 생성자 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(NorthWestMovementTest): 불필요한 접근제어자 제거

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style: 불필요한 개행 삭제

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor: 디미터 규칙 준수

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style: 가독성을 위해 개행 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
refactor(Position): 불필요한 메서드 삭제

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
docs: 개선 필요 사항 분리

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

hyxrxn committed:
style(Position): 가독성을 위해 this 추가

Co-authored-by: leegwichan <cnddkscndgus@naver.com>

woonjangahn commented on README.md:
```diff
@@ -1,7 +1,39 @@
+### 출력
+- [x] 체스판에서 각 진영은 검은색(대문자)과 흰색(소문자) 편으로 구분한다.
+
+## 개선 필요 사항
+- [ ] 마지막 위치에 적 말이 있을 경우, 잡아먹는다.
+- [ ] 흰색부터 번갈아가며 플레이한다.
```
"체스 말의 이동 규칙을 찾아보고 체스 말이 이동할 수 있도록 구현한다."에 사실 아토가 명시해준 개선 필요 사항이 포함되어있는 것 같은데요, 이번 단계 안에 구현 부탁드릴게요.

woonjangahn replied:
그리고 잘못된 커맨드 입력시에 프로세스가 종료되는 부분도 수정 부탁드릴게요 🙇

woonjangahn commented on src/main/java/chess/domain/BoardFactory.java:
```diff
@@ -0,0 +1,74 @@
+
+public class BoardFactory {
+
+    private BoardFactory() {
+
+    }
```
불필요한 공백줄은 삭제해주세요.

woonjangahn commented on src/main/java/chess/domain/BoardFactory.java:
```diff
@@ -0,0 +1,74 @@
+
+    }
+
+    public static Board startGame() {
```
팩토리 패턴을 사용해주셨네요 👍

woonjangahn replied:
다만, 그렇다면 startGame보다는 Board를 반환하는 느낌의 명시적인 네이밍을 해주면 좋겠어요.

woonjangahn commented on src/main/java/chess/domain/piece/Bishop.java:
```diff
@@ -0,0 +1,18 @@
+import chess.domain.movement.continuous.SouthWestMovement;
+import java.util.List;
+
+public final class Bishop extends Piece {
```
final 같은 modifier은 일관성 있게 사용해주세요.

woonjangahn commented on src/main/java/chess/domain/position/Position.java:
```diff
@@ -0,0 +1,66 @@
+        return this.rank == rank;
+    }
+
+    public int calculateFileDifference(Position position) {
+        return this.file.calculateDifference(position.file);
+    }
+
+    public int calculateRankDifference(Position position) {
+        return this.rank.calculateDifference(position.rank);
+    }
```
같은 클래스끼리 비교할 때 other같은 변수명을 쓰면 조금 더 가독성이 좋아집니다.

woonjangahn commented on src/main/java/chess/dto/PieceDto.java:
```diff
@@ -0,0 +1,13 @@
+
+public record PieceDto(PieceType type, boolean isBlack) {
+
+    public static PieceDto from(Piece piece) {
```
record의 주생성자가 public이라 생성시에 주생성자를 사용할지 팩토리 메서드를 사용할지 헷갈릴 수 있을 것 같아요.

woonjangahn commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,71 @@
+
+    private void printBoardOneLine(Map<Position, PieceDto> board, Rank rank) {
+        for (File file : FILE_ORDER) {
+            PieceDto piece = board.get(new Position(file, rank));
```
board도 nullable보다는 Optional을 사용해보는것은 어떨까요?

woonjangahn reviewed (changes requested):
안녕하세요 아토 👋 

체스 1, 2단계 구현 고생하셨어요 👍 

코멘트 확인 부탁드려요 🙇

hyxrxn committed:
refactor(BoardFactory): 메서드명 변경

hyxrxn committed:
refactor(PieceDto): 클래스로 변경

- 주 생성자를 private으로 변경

hyxrxn committed:
feat(BoardDto): 체스 보드 출력을 위한 Dto 생성

- OutputView에서 null 사용 제거
- ChessGame의 역할 분리

hyxrxn committed:
feat(Board): 팀 규칙 추가

- 흰 말부터 차례로 이동
- 같은 팀 말이 있는 위치로 이동 불가

hyxrxn committed:
refactor(MovementRule): 상속되지 않을 클래스에 final 추가

hyxrxn committed:
refactor(Position): 파라미터 변수명 변경

hyxrxn committed:
docs: 기능 요구 사항 추가

hyxrxn committed:
feat(MovementRule): 적 여부 파라미터에 추가

- 폰 이동 시 적 여부 판단해서 이동 가능 여부 반환
- 폰 대각선 이동 방법 추가

hyxrxn committed:
feat(MovementRule): 도착 위치는 이동 경로에 포함되지 않도록 변경

- 이동 경로란 출발 위치에서 도착 위치까지 갈 때 거쳐가는 칸들
- 출발 위치에서 시작해 이동 경로를 지나 도착 위치에 도달
- 출발 위치에서 도착 위치 사이에 거쳐가는 칸이 없는 경우 빈 리스트 반환

hyxrxn committed:
feat(ChessGame): 잘못된 입력 시 오류 메시지 출력 후 재입력

hyxrxn committed:
style(ChessGame): 컨벤션에 맞춰 공백 추가

hyxrxn committed:
refactor(File): 오타 수정

hyxrxn committed:
refactor(Board): 메서드 간략화

woonjangahn commented on src/main/java/chess/domain/position/File.java:
```diff
@@ -0,0 +1,53 @@
+
+    private final int westToEast;
+
+    File(int westToEast) {
```
westToEast가 File의 값을 나타내는 변수명으로는 모호한것 같아요.

woonjangahn commented on src/main/java/chess/domain/position/Position.java:
```diff
@@ -0,0 +1,66 @@
+        return this.rank.calculateDifference(other.rank);
+    }
+
+    public Position moveToEast() {
+        return new Position(file.toEast(), rank);
+    }
+
+    public Position moveToWest() {
+        return new Position(file.toWest(), rank);
+    }
+
+    public Position moveToNorth() {
+        return new Position(file, rank.toNorth());
+    }
+
+    public Position moveToSouth() {
+        return new Position(file, rank.toSouth());
+    }
+
```
이렇게 만들어서 메소드 체이닝을 한 것이 굉장히 깔끔해서 맘에 드네요 👍

woonjangahn reviewed (approved):
안녕하세요 아토 👋 

추가 요구사항 구현해주시느라 고생많으셨어요.
코드가 굉장히 깔끔하네요 💯 

다음 단계에서 뵙겠습니다 🙇
